File: lua/src/02-2.lua
local input = io.open('../data/02.txt'):read('*a')

local total = 0

for line in input:gmatch ('[^\n]+') do
  local num, batches = line:match 'Game (%d+): (.*)'
  num = tonumber(num)
  local maxes = { red = 0, green = 0, blue = 0 }
  for batch in batches:gmatch '[^;]+' do
    for item in batch:gmatch '[^,]+' do
      local count, color = item:match '(%d+) (%w+)'
      count = tonumber(count)
      if count > maxes [color] then
        maxes[color] = count
      end
    end
  end
  total = total + maxes.red * maxes.green * maxes.blue
end

print (total)

File: typescript/src/02-2.ts
import { readFileSync } from 'fs';

const input = readFileSync ('../data/02.txt', 'utf-8')
    .trimEnd(); // remove trailing newline

let total = 0
input.split('\n').forEach(line => {
    const match = line.match(/Game (\d+): (.*)/);
    if (!match) { throw new Error(); }
    const batches = match[2];
    const maxes: { [key: string]: number } = { red: 0, green: 0, blue: 0 };

    batches.split('; ').forEach(batch => {
        batch.split(', ').forEach(item => {
            let itemMatch = item.match(/(\d+) (\w+)/)
            if (!itemMatch) { throw new Error(); }
            const count = parseInt(itemMatch[1]);
            const color = itemMatch[2];
            maxes [color] = Math.max(count, maxes[color]);
        });
    });

    total += maxes.red * maxes.green * maxes.blue;
});
console.log (total)
